Tasks/Common/VstsAzureRestHelpers_/VstsAzureRestHelpers_.psm1

```diff
@@ -258,28 +281,128 @@ function Has-ObjectProperty {
     }
 }
 
+# Get access token - Main Point
 function Get-AzureRMAccessToken {
     [CmdletBinding()]
     param(
         [Parameter(Mandatory = $true)] $endpoint,
-        [parameter(Mandatory = $false)] $overrideResourceType = $null
-    ) 
+        [parameter(Mandatory = $false)] $overrideResourceType = $null,
+        [parameter(Mandatory = $false)] $useMSAL = $true
+    )
+
+    $accessToken = @{
+        token_type = $null
+        access_token = $null
+        expires_on = $null
+    }
+
+    Write-Verbose "Get-AzureRMAccessToken - started - endpoint=$endpoint - scheme=$($endpoint.Auth.Scheme)"
+    Write-Verbose "MSAL - useMSAL = $useMSAL"
 
+    # ManagedIdentity - access token
     if ($endpoint.Auth.Scheme -eq $MsiConnection ) {
-        Get-MsiAccessToken -endpoint $endpoint -retryCount 0 -timeToWait 0 -overrideResourceType $overrideResourceType
+        Write-Verbose "MSAL - ManagedIdentity is used"
+        $accessToken = Get-MsiAccessToken -endpoint $endpoint -retryCount 0 -timeToWait 0 -overrideResourceType $overrideResourceType
     }
+    # MSAL - access token
+    elseif ($useMSAL) {
+        $result = Get-AccessTokenMSAL -endpoint $endpoint -overrideResourceType $overrideResourceType
+
+        $accessToken.token_type = $result.TokenType
+        $accessToken.access_token = $result.AccessToken
+        $accessToken.expires_on = $result.ExpiresOn.ToUnixTimeSeconds()
+    }
+    # ADAL - access token - @DEPRECATED - will be removed
     else {
         if ($Endpoint.Auth.Parameters.AuthenticationType -eq 'SPNCertificate') {
-            Get-SpnAccessTokenUsingCertificate -endpoint $endpoint -overrideResourceType $overrideResourceType
+            $accessToken = Get-SpnAccessTokenUsingCertificate -endpoint $endpoint -overrideResourceType $overrideResourceType
         }
         else {
-            Get-SpnAccessToken -endpoint $endpoint -overrideResourceType $overrideResourceType
+            $accessToken = Get-SpnAccessToken -endpoint $endpoint -overrideResourceType $overrideResourceType
         }
-    }   
+    }
+
+    return $accessToken;
+}
+
+# Build MSAL instance
+function Build-MSAL {
```

Comment:
Maybe return the MSAL client in this method instead of relying on a global variable everywhere? By all means, use a global variable within the method to cache the client so you don't have to recreate it; but I think this should just be an implementation detail of this method, and users of this method should not have to rely on it.

Reply:
I'm planning to create and use `Get-MSAL` which will be more convenient in this case.

Reply:
I've changed to structure. `Get-MSALInstance` will be used. It will control whether if instance is already created or not.